### ISSUE TITLE
array_t assignment operators

### ISSUE THREAD

**Issue by uentity** [pybind/pybind11#725]
Hi!
I'm in process of writing type converter for my own array class based on @jagerman approach to bind Eigen types. Let's consider avoid-copy `type_caster::load` function for `Eigen::Ref< MatrixType>` which does in general the following things:
1) construct temp array `aref` from source handle;
2) saves `aref` into member variable copy_or_ref by using either
`copy_or_ref = std::move(aref)` for no-copy case or
`copy_or_ref = aref` if copy of source array is required;
3) construct Eigen target type from `copy_or_ref` and return it.

After I copy-pasted above implementation to my own type caster I discovered that step 2 break things. I pass simple numpy vector of size 10 into my test function. And no matter do we need copy or not after assignment:
1) `copy_or_ref` always has size = 1;
2) `copy_or_ref.data()` differs from `aref.data()`;
3) first element of `copy_or_ref` is a copy of first element from `aref`.
It seems like assignment creates new array of size = 1 and copies the first element from source array.

If I eliminate assignment operators from my type caster and create `array_t` copies using copy constructor only, then everything works as expected!

What `array_t` assignment operator supposed to do? Should it make different things compared to copy constructor?

**Comment by jagerman** [comment]
It's a bit hard to say without the specific code, but I think the issue in Step 2 is:

`copy_or_ref = aref;` isn't quite what the eigen code does: instead it first does an `Array copy = Array::ensure(src)` which will do the actual coercion to a numpy array (copying and/or type converting as needed), then assigns that to `copy_or_ref`.  So it's really the `ensure` that makes a copy here.

**Comment by jagerman** [comment]
(Also there should have been a `std::move` for the copy code path as well--I just added it).

**Comment by jagerman** [comment]
So bigger picture, the `py::array` and `py::array_t<Type>` (and likewise for most of the python type wrappers pybind provides) are *not* direct storage of arrays; it's probably better to think of them as shared pointers to the underlying python object.  Thus a copy or a move is really just copying or moving the wrapper/shared pointer, not the object itself: the object itself is owned and managed by Python.

**Comment by uentity** [comment]
@jagerman sorry, I messed up with typedefs and got wrong type  for `aref` and `copy_or_ref` (actually it was my wrapper derived from `array_t` with overloaded assignment operator). Thanks for your patience! I checked that for pure `array_t` assignment works as it have to. And I clearly understand what you've said above, just faced this strange issue )
Thanks one more time for great code!
Think this should be closed.
